File: internal/context/cursorrules.go
package context

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v3"
)

// cursorRuleFrontmatter mirrors the subset of Cursor's .mdc frontmatter
// fields relevant to deciding whether a rule is unconditionally applicable
// in a whole-PR review, where there is no single-file diff-aware matching
// and no interactive chat session for @-mentions to happen in.
type cursorRuleFrontmatter struct {
	Description string `yaml:"description"`
	Globs       string `yaml:"globs"`
	AlwaysApply *bool  `yaml:"alwaysApply"`
}

// parseMDCFrontmatter splits an .mdc file into its frontmatter and body. ok
// is false when the file has no valid "---" delimited frontmatter block, in
// which case the caller should skip it.
func parseMDCFrontmatter(data []byte) (fm cursorRuleFrontmatter, body string, ok bool) {
	s := string(data)
	if !strings.HasPrefix(s, "---\n") && !strings.HasPrefix(s, "---\r\n") {
		return cursorRuleFrontmatter{}, "", false
	}

	rest := s[strings.IndexByte(s, '\n')+1:]
	end := strings.Index(rest, "\n---\n")
	sepLen := len("\n---\n")
	if end == -1 {
		end = strings.Index(rest, "\n---\r\n")
		sepLen = len("\n---\r\n")
	}
	if end == -1 {
		return cursorRuleFrontmatter{}, "", false
	}

	raw := rest[:end]
	remainder := rest[end+sepLen:]

	if err := yaml.Unmarshal([]byte(raw), &fm); err != nil {
		return cursorRuleFrontmatter{}, "", false
	}
	return fm, strings.TrimSpace(remainder), true
}

// isUnconditionallyApplicable reports whether a rule should be folded into
// a whole-PR review's project context: alwaysApply: true, or rules with no
// file-scoping (globs) at all. Rules scoped to specific files via globs are
// skipped, since a whole-PR review has no single file to match them against.
func isUnconditionallyApplicable(fm cursorRuleFrontmatter) bool {
	if fm.AlwaysApply != nil {
		return *fm.AlwaysApply
	}
	return strings.TrimSpace(fm.Globs) == ""
}

// DiscoverCursorContext reads AGENTS.md and applicable .cursor/rules/*.mdc
// files from repoRoot (root level only, not recursive) and returns their
// concatenated content formatted as "### <name>" sections, consistent with
// how project context_files are rendered.
func DiscoverCursorContext(repoRoot string) string {
	var sb strings.Builder

	if data, err := os.ReadFile(filepath.Join(repoRoot, "AGENTS.md")); err == nil {
		if content := strings.TrimSpace(string(data)); content != "" {
			sb.WriteString("### AGENTS.md\n")
			sb.WriteString(content)
			sb.WriteString("\n\n")
		}
	}

	rulesDir := filepath.Join(repoRoot, ".cursor", "rules")
	entries, err := os.ReadDir(rulesDir)
	if err != nil {
		return sb.String()
	}

	for _, e := range entries {
		if e.IsDir() || filepath.Ext(e.Name()) != ".mdc" {
			continue
		}
		data, err := os.ReadFile(filepath.Join(rulesDir, e.Name()))
		if err != nil {
			continue
		}
		fm, body, ok := parseMDCFrontmatter(data)
		if !ok || body == "" || !isUnconditionallyApplicable(fm) {
			continue
		}
		fmt.Fprintf(&sb, "### .cursor/rules/%s\n%s\n\n", e.Name(), body)
	}

	return sb.String()
}

File: src/context/cursor-rules.ts
import * as fs from 'fs';
import * as path from 'path';
import { parse as parseYaml } from 'yaml';

interface CursorRuleFrontmatter {
  description?: string;
  globs?: string;
  alwaysApply?: boolean;
}

export function parseMDCFrontmatter(data: string): { fm: CursorRuleFrontmatter; body: string; ok: boolean } {
  if (!data.startsWith('---\n') && !data.startsWith('---\r\n')) {
    return { fm: {}, body: '', ok: false };
  }

  const firstNewline = data.indexOf('\n');
  const rest = data.slice(firstNewline + 1);

  let end = rest.indexOf('\n---\n');
  let sepLen = '\n---\n'.length;
  if (end === -1) {
    end = rest.indexOf('\n---\r\n');
    sepLen = '\n---\r\n'.length;
  }
  if (end === -1) {
    return { fm: {}, body: '', ok: false };
  }

  const raw = rest.slice(0, end);
  const remainder = rest.slice(end + sepLen);

  try {
    const fm = parseYaml(raw) as CursorRuleFrontmatter;
    return { fm: fm ?? {}, body: remainder.trim(), ok: true };
  } catch {
    return { fm: {}, body: '', ok: false };
  }
}

export function isUnconditionallyApplicable(fm: CursorRuleFrontmatter): boolean {
  if (fm.alwaysApply !== undefined) {
    return fm.alwaysApply;
  }
  return !(fm.globs ?? '').trim();
}

export function loadCursorContext(repoRoot: string): string {
  let result = '';

  try {
    const agentsPath = path.join(repoRoot, 'AGENTS.md');
    const content = fs.readFileSync(agentsPath, 'utf-8').trim();
    if (content) {
      result += `### AGENTS.md\n${content}\n\n`;
    }
  } catch {
    // file not found, skip
  }

  const rulesDir = path.join(repoRoot, '.cursor', 'rules');
  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(rulesDir, { withFileTypes: true });
  } catch {
    return result;
  }

  for (const entry of entries) {
    if (entry.isDirectory() || path.extname(entry.name) !== '.mdc') continue;

    let data: string;
    try {
      data = fs.readFileSync(path.join(rulesDir, entry.name), 'utf-8');
    } catch {
      continue;
    }

    const { fm, body, ok } = parseMDCFrontmatter(data);
    if (!ok || !body || !isUnconditionallyApplicable(fm)) continue;

    result += `### .cursor/rules/${entry.name}\n${body}\n\n`;
  }

  return result;
}
